datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -97,7 +97,18 @@ fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
     let new_exprs = plan
         .expressions()
         .into_iter()
-        .map(|expr| expr.rewrite(&mut expr_rewriter))
+        .map(|expr| {
+            let original_name = expr.name()?;
```

Comment:
I think the root cause issue is that the `UnwrapCastInComparison` can add potentially change the expression but it doesn't add an alias so the output name changes

Bigger picture there are at least three places that we have rediscovered this same problem when rewriting expressions --https://github.com/apache/arrow-datafusion/issues/3555  and https://github.com/apache/arrow-datafusion/blob/master/datafusion/optimizer/src/simplify_expressions.rs#L316  I will try and make a follow on PR to clean them all up. In particular, I think this is something `from_plan` could potentially handle

Reply:
`expr.name()` here hides the casts that are added by this optimization rule, so `expr.name()` is the same as the original name (even though the expression is now different), and the alias does not get added.

Reply:
I think this optimizer rule may actually change the name (e.g. from `Int64(0)` to `Int32(0)`) which i think is the root cause of the issue in this bug